drivers/video/video_sw_generator.c

```diff
@@ -261,6 +270,66 @@ static inline int video_sw_generator_set_ctrl(const struct device *dev, unsigned
 	return 0;
 }
 
+static int video_sw_generator_set_frmival(const struct device *dev, enum video_endpoint_id ep,
+					  struct video_frmival *frmival)
+{
+	struct video_sw_generator_data *data = dev->data;
+
+	if (frmival->denominator && frmival->numerator) {
+		data->frame_rate = MIN(DIV_ROUND_CLOSEST(frmival->denominator, frmival->numerator),
+				       MAX_FRAME_RATE);
+	} else {
+		return -EINVAL;
+	}
+
+	frmival->numerator = 1;
+	frmival->denominator = data->frame_rate;
+
+	return 0;
+}
+
+static int video_sw_generator_get_frmival(const struct device *dev, enum video_endpoint_id ep,
+					  struct video_frmival *frmival)
+{
+	struct video_sw_generator_data *data = dev->data;
+
+	frmival->numerator = 1;
+	frmival->denominator = data->frame_rate;
+
+	return 0;
+}
+
+static int video_sw_generator_enum_frmival(const struct device *dev, enum video_endpoint_id ep,
+					   struct video_frmival_enum *fie)
+{
+	int i = 0;
+
+	if (ep != VIDEO_EP_OUT || fie->index) {
+		return -EINVAL;
+	}
+
+	while (fmts[i].pixelformat && (fmts[i].pixelformat != fie->pixelformat)) {
+		i++;
+	}
+
+	if ((i == ARRAY_SIZE(fmts)) || (fie->width > fmts[i].width_max) ||
+	    (fie->width < fmts[i].width_min) || (fie->height > fmts[i].height_max) ||
+	    (fie->height < fmts[i].height_min)) {
+		return -EINVAL;
+	}
```

Comment:
FYI: there is an opportunity to reduce the drivers code size by moving these to [`drivers/video.h`](https://github.com/zephyrproject-rtos/zephyr/pull/72254/files#diff-70c4de6949b97c13ac893b4bfb16fbcc433593fbfa96ec8947bb5cb83c4d04bcR485)

Reply:
How do you get "fmts" in video.h ?

Reply:
You are right! Apologies I missed that there was no width/height as part of `fie`.
Each driver might encode that differently.
It could be added to something like `video_sensor_skeleton.c` (if that ever happens).
Sorry for the noise!